daemon/images/images_test.go

```diff
@@ -0,0 +1,102 @@
+
+type mockLayerGetReleaser struct{}
+
+func (ls *mockLayerGetReleaser) Get(layer.ChainID) (layer.Layer, error) {
+	return nil, nil
+}
+
+func (ls *mockLayerGetReleaser) Release(layer.Layer) ([]layer.Metadata, error) {
+	return nil, nil
+}
+
+func defaultFSStoreBackend(t *testing.T) (image.StoreBackend, func()) {
+	tmpdir, err := ioutil.TempDir("", "images-fs-store")
+	assert.NilError(t, err)
+
+	fsBackend, err := image.NewFSStoreBackend(tmpdir)
+	assert.NilError(t, err)
+
+	return fsBackend, func() { os.RemoveAll(tmpdir) }
+}
+
+func defaultImageStore(t *testing.T) (image.Store, func()) {
+	fsBackend, cleanup := defaultFSStoreBackend(t)
+
+	mlgrMap := make(map[string]image.LayerGetReleaser)
+	mlgrMap[runtime.GOOS] = &mockLayerGetReleaser{}
+	store, err := image.NewImageStore(fsBackend, mlgrMap)
+	assert.NilError(t, err)
+
+	return store, cleanup
+}
+
+func defaultReferenceStore(t *testing.T, loadFrom []byte) (reference.Store, *os.File) {
+	jsonFile, err := ioutil.TempFile("", "tag-store-test")
+	assert.NilError(t, err)
+
+	defer os.RemoveAll(jsonFile.Name())
+
+	// Write canned json to the temp file
+	_, err = jsonFile.Write(loadFrom)
+	assert.NilError(t, err)
+
+	rs, err := reference.NewReferenceStore(jsonFile.Name())
+	assert.NilError(t, err)
+
+	return rs, jsonFile
+
+}
+
+func TestFilterMultipleImageReferences(t *testing.T) {
+	// imageStore initialize
+	is, cleanup := defaultImageStore(t)
+	defer cleanup()
+
+	_, err := is.Create([]byte(`{"comment": "test1", "rootfs": {"type": "layers"}, "id": "1"}`))
+	assert.NilError(t, err)
+
+	_, err = is.Create([]byte(`{"comment": "test2", "rootfs": {"type": "layers"}, "id": "2"}`))
+	assert.NilError(t, err)
+
+	_, err = is.Create([]byte(`{"comment": "test3", "rootfs": {"type": "layers"}, "id": "3"}`))
+	assert.NilError(t, err)
+
+	// Reference store initialize
+	rs, jf := defaultReferenceStore(t, marshalledSaveLoadTestCases)
+	jf.Close()
+
+	daemon := &ImageService{
+		imageStore:     is,
+		referenceStore: rs,
+	}
+
+	var images []*types.ImageSummary
+	expectedImageCount := 10
+
+	for i := 0; i < 10; i++ {
+		ifs := filters.NewArgs(filters.Arg("reference", "busybox"), filters.Arg("reference", "ls"))
+		img, err := daemon.Images(ifs, false, false)
+		assert.NilError(t, err)
+
+		images = append(images, img...)
+	}
+
+	assert.Equal(t, expectedImageCount, len(images))
```

Comment:
How is this testing the original bug? It seems this would always be length of 10 because of how `images` is created (in a loop).

Reply:
No, if you remove the patch that I added to the code, it'll not return 10 anymore (will be a random number between 1 and 10). So it is a test for that bug.